### PR TITLE
IMS-51 PDF Exporter

### PR DESCRIPTION
**Pull request by Rexmoon**
### Related ticket

Ticket: [IMS-51](https://developer-solutions.atlassian.net/browse/IMS-51)

### Description

- Export charts feat added

### Medias (if needed)

Simulator Screen Recording - iPad (10th generation) - 2024-08-12 at 00 39 07

### Testplan

- [ ] I tested this change on my local environment.
- [ ] I ran the project using multiple simulators in Xcode.
- [ ] I ran the project using Real device (if applicable).
- [ ] I ran the unit tests.
- [ ] I requested this endpoint on Postman (if applicable).
- [ ] I requested this endpoint on Swagger documentation (if applicable).

### Checklist

- [ ] I added tests for this change.
- [ ] I checked the code style and run the linter.
- [ ] I added necessary documentation (if applicable).

### CODE REVIEW

File: ims-apple/IMS Apple/Utilities/Exporter/PDFExporter.swift [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                    }
+                    .padding()
+                }
+        }
+    }
+    
+    func render() -> URL {
+        let rendered = ImageRenderer(content: content)
+        
+        let url: URL = .documentsDirectory.appendingPathComponent("Charts.pdf", conformingTo: .pdf)
```
Comment (derian-cordoba) [inline ims-apple/IMS Apple/Utilities/Exporter/PDFExporter.swift]: Can you add a variable to define the PDF title pls?

Comment (Rexmoon) [reply ims-apple/IMS Apple/Utilities/Exporter/PDFExporter.swift]: Rigth 👍

File: ims-apple/IMS Apple/Views/Graph/GraphView.swift [hunk 2]
```diff
@@ -19,15 +19,17 @@ private enum Constants {
...
+            Grid(horizontalSpacing: Constants.spacing,
+                 verticalSpacing: Constants.spacing) {
+                firstGridRow
+                
+                secondGridRow
+            }
+            .padding()
+            .background(.grayBackground)
+            .frame(maxWidth: .infinity, maxHeight: .infinity)
+        })
```
Comment (santelizvargas) [inline ims-apple/IMS Apple/Views/Graph/GraphView.swift]: We dont need a container for this implementation

File: ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            content
+                .contextMenu {
+                    HStack {
+                        #if os(iOS)
+                        ShareLink("Exportar", item: render())
+                        #else
+                        Button("Exportar") {
+                            NSWorkspace.shared.selectFile(nil, inFileViewerRootedAtPath: render().path)
+                        }
+                        #endif
```
Comment (santelizvargas) [inline ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift]: We need to add some view for use the OSType enum

File: ims-apple/IMS Apple/Views/Graph/GraphView.swift (hunk 2 above)
Comment (Rexmoon) [reply ims-apple/IMS Apple/Views/Graph/GraphView.swift]: What do you propose?

File: ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift (hunk 3 above)
Comment (Rexmoon) [reply ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift]: Enum represents a platform? 🤔

Comment (santelizvargas) [reply ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift]: Yes

File: ims-apple/IMS Apple/Views/Graph/GraphView.swift (hunk 2 above)
Comment (santelizvargas) [reply ims-apple/IMS Apple/Views/Graph/GraphView.swift]: ViewModifier

https://developer.apple.com/documentation/swiftui/viewmodifier

Comment (santelizvargas) [reply ims-apple/IMS Apple/Views/Graph/GraphView.swift]: Or only an extension

File: ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift (hunk 3 above)
Comment (Rexmoon) [reply ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift]: Copy that!

File: ims-apple/IMS Apple/Views/Graph/GraphView.swift (hunk 2 above)
Comment (Rexmoon) [reply ims-apple/IMS Apple/Views/Graph/GraphView.swift]: Cool, thanks I will working on it.

File: ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+//  IMS Apple
+//
+//  Created by Brandon Santeliz on 8/13/24.
+//
+
+import SwiftUI
+
+extension View {
+    
+    /// Make a PDF file from View content
```
Comment (santelizvargas) [inline ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift]: Remove this comment or move to the top

File: ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    private let content: Content
+    
+    init(fileName: String, content: () -> Content) {
+        self.fileName = fileName
+        self.content = content()
+    }
+    
+    var body: some View {
+        
+        let url: URL = render()
```
Comment (santelizvargas) [inline ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift]: Remove var and use directly in the switch

File: ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+    
+    func render() -> URL {
+        let rendered = ImageRenderer(content: content)
+        
+        let url: URL = .documentsDirectory.appendingPathComponent("\(fileName).pdf", conformingTo: .pdf)
+        
+        rendered.render { size, context in
+            var box = CGRect(x: 0, y: 0, width: size.width, height: size.height)
```
Comment (santelizvargas) [inline ims-apple/IMS Apple/Utilities/Exporter/PDFExporterContainer.swift]:
```suggestion
            var box = CGRect(x: .zero, y: .zero, width: size.width, height: size.height)
```

File: ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    func exportView(with fileName: String) -> some View {
+        PDFExporterContainer(
+            fileName: fileName) {
+                self
+            }
+    }
```
Comment (santelizvargas) [inline ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift]: Container no needed

File: ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift (hunk 4 above)
Comment (Rexmoon) [reply ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift]: Is a function documentation

File: ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift (hunk 7 above)
Comment (Rexmoon) [reply ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift]: Render function need View component instead of Content of view to work.

Comment (santelizvargas) [reply ims-apple/IMS Apple/Utilities/Extensions/View+PDFExporter.swift]: Add extension or View Modifier

File: ims-apple/IMS Apple/Utilities/Components/Buttons/ExportPDFButton.swift [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    
+    private var labelView: some View {
+        HStack {
+            Image(systemName: "square.and.arrow.up.fill")
+            Text("Exportar")
+        }
+        .foregroundStyle(.white)
+    }
+    
+    // MARK: - Current Directory
```
Comment (Rexmoon) [inline ims-apple/IMS Apple/Utilities/Components/Buttons/ExportPDFButton.swift:48]: Is neccessary a MARK by variable?

File: ims-apple/IMS Apple/Utilities/Exporter/FileExporter.swift [hunk 9]
```diff
@@ -15,7 +15,10 @@ final class FileExporter {
     private lazy var fileFactory: FileFactory = FileFactory()
     
     private lazy var downloadsDirectoryUrl: URL = {
-        FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask)[0]
+        guard let url = FileManager.default.urls(for: .downloadsDirectory, in: .userDomainMask).first else {
+            fatalError("Fatal Error: File Urls list is empty")
+        }
+        return url
```
Comment (Rexmoon) [inline ims-apple/IMS Apple/Utilities/Exporter/FileExporter.swift]:
```swift
private lazy var downloadsDirectoryUrl: URL = {
        .downloadsDirectory
    }()
```

**Review by santelizvargas — APPROVED** [review]
LGTM 👍

**Pull request merged**